contracts/compounder/GenericVault.sol

```diff
@@ -0,0 +1,350 @@
+            IBasicRewards(extraRewards[i]).stake(to, amount);
+        }
+    }
+
+    /* --------------------------------------------------------------
+     * EIP-4626 functions
+    ----------------------------------------------------------------- */
+
+    /// @notice The address of the underlying token used for the Vault for
+    /// accounting, depositing, and withdrawing.
+    function asset() public view returns (address) {
+        return underlying;
+    }
+
+    /// @notice Total amount of the underlying asset that is “managed” by Vault.
+    function totalAssets() public view returns (uint256) {
+        return totalUnderlying();
+    }
+
+    /// @notice The amount of shares that the Vault would exchange for the amount
+    /// of assets provided, in an ideal scenario where all the conditions are met.
+    function convertToShares(uint256 _assets) public view returns (uint256) {
+        return totalSupply() == 0 ? _assets : (_assets * totalSupply()) / totalAssets();
+    }
+
+    /// @notice The amount of assets that the Vault would exchange for the amount
+    /// of shares provided, in an ideal scenario where all the conditions are met.
+    function convertToAssets(uint256 _shares) public view returns (uint256) {
+        return totalSupply() == 0 ? _shares : (_shares * totalUnderlying()) / totalSupply();
+    }
+
+    /// @notice Maximum amount of the underlying asset that can be deposited into
+    /// the Vault for the receiver, through a deposit call.
+    function maxDeposit(address) public pure returns (uint256) {
+        return type(uint256).max;
+    }
+
+    /// @notice Allows an on-chain or off-chain user to simulate the effects of
+    /// their deposit at the current block, given current on-chain conditions.
+    function previewDeposit(uint256 _assets) public view returns (uint256) {
+        return convertToShares(_assets);
+    }
+
+    /// @notice Maximum amount of shares that can be minted from the Vault
+    /// for the receiver, through a mint call.
+    function maxMint(address) public pure returns (uint256) {
+        return type(uint256).max;
+    }
+
+    /// @notice Allows an on-chain or off-chain user to simulate the effects of
+    /// their mint at the current block, given current on-chain conditions.
+    function previewMint(uint256 _shares) public view returns (uint256) {
+        return convertToAssets(_shares);
+    }
+
+    /// @notice Mints exactly shares Vault shares to receiver by depositing
+    /// assets of underlying tokens.
+    function mint(uint256 _shares, address _receiver) public returns (uint256) {
+        uint256 assets = convertToAssets(_shares);
+        return deposit(assets, _receiver);
+    }
+
+    /// @notice Maximum amount of the underlying asset that can be withdrawn
+    /// from the owner balance in the Vault, through a withdraw call.
+    function maxWithdraw(address _owner) public view returns (uint256) {
+        return convertToAssets(balanceOf(_owner));
```

Comment:
```suggestion
        return previewRedeem(maxRedeem((_owner)));
```

The value currently returned isn't quite right and doesn't account for fees, should we do this? 

You can't use the value with the withdraw() function as the value returned currently exceeds user balances.